lib/nerves_runtime/log_tailer.ex

```diff
@@ -0,0 +1,137 @@
+defmodule Nerves.Runtime.LogTailer do
+  use GenServer
+
+  require Logger
+
+  defp gen_server_name(:syslog), do: __MODULE__.Syslog
+  defp gen_server_name(:kmsg), do: __MODULE__.Kmsg
+
+  @doc """
+  `type` must be `:syslog` or `:kmsg` to indicate which log to tail with this
+  process. They're manages by separate processes both to isolate failures and
+  to simplify the handling of messages being sent back from the ports.
+  """
+  @spec start_link(:syslog | :kmsg) :: {:ok, pid()}
+  def start_link(type) do
+    GenServer.start_link(__MODULE__, type, name: gen_server_name(type))
+  end
+
+  @spec init(:syslog | :kmsg) :: {:ok, %{type: atom(), port: port(), buffer: binary()}}
+  def init(type) do
+    if Application.get_env(:nerves_runtime, :disable_syslog, false) do
+      :ignore
+    else
+      port =
+        type
+        |> port_binary_name()
+        |> open_port()
+
+      {:ok, %{type: type, port: port, buffer: ""}}
+    end
+  end
+
+  def handle_info({port, {:data, {:noeol, fragment}}}, %{port: port, buffer: buffer} = state) do
+    {:noreply, %{state | buffer: buffer <> fragment}}
+  end
+
+  def handle_info({port, {:data, {:eol, fragment}}}, %{type: type, port: port, buffer: buffer} = state) do
+    handle_message(type, buffer <> fragment)
+    {:noreply, %{state | buffer: ""}}
+  end
+
+  defp port_binary_name(:syslog), do: "log_tailer"
+  defp port_binary_name(:kmsg), do: "kmsg_tailer"
+
+  defp open_port(name) do
+    Port.open({:spawn_executable, executable(name)}, [
+      {:line, 1024},
+      :use_stdio,
+      :binary,
+      :exit_status
+    ])
+  end
+
+  defp executable(name) do
+    :nerves_runtime
+    |> :code.priv_dir()
+    |> Path.join(name)
+  end
+
+  defp handle_message(type, data) do
+    case parse_syslog_message(data) do
+      %{facility: facility, severity: severity, message: message} ->
+        Logger.bare_log(
+          logger_level(severity),
+          message,
+          [module: gen_server_name(type), facility: facility, severity: severity]
+        )
+
+      _ ->
+        # This is unlikely to ever happen, but if a message was somehow
+        # malformed and we couldn't parse the syslog priority, we should
+        # still do a best-effort to pass along the raw data.
+        Logger.info(data, module: gen_server_name(type))
```

Comment:
The values that `gen_server_name` return aren't modules that exist anywhere with code.  Might be confusing to some.

Reply:
Yeah, I thought about that too, but at the end of the day, we're kind of pretending that there are two different modules, except without actually having two modules because they'd end up being the same.

I just figured that there was value in having a simple way to differentiate where the logs were coming from, in a way people would expect, even if the module names aren't actual Elixir modules as far as the compiler is concerned. For example, instead of using my own arbitrary metadata tag (like `type: :kmsg`), I am using an existing construct that people familiar with `Logger` metadata filtering will already be comfortable with. I doubt anyone would really care that the module doesn't exist - they just want to have a way to filter messages in or out of a given `Logger` back-end.

Having said all that, maybe the right answer is just to just use the real module name (`Nerves.Runtime.LogTailer`) and either not differentiate between `syslog` and `kmsg` sources, or additionally include the `:type` metadata as well. This also kind of begs the question about whether we really need to side-step the default `Logger` metadata by using `Logger.bare_log/3`. What makes this use of `Logger` special where that divergence is meaningful?

I don't really have strong opinions either way, but I think I'm leaning more toward including the default module/file/line metadata just because it's normally there (the `Logger` docs say it's going to be there), and it's really an end-developer concern about whether they want to display them in the back-end they're using. I don't see a compelling reason to strip that metadata out, for instance, compared to the other uses of `Logger` in the rest of the Nerves ecosystem.

Thoughts?